build_runner/lib/src/asset/finalized_reader.dart

```diff
@@ -19,6 +21,20 @@ class FinalizedReader implements AssetReader {
     this._assetGraph,
   );
 
+  /// Returns a reason why [id] is not readable, or null if it is readable.
+  Future<UnreadableAssetReason> canReadReason(AssetId id) async {
+    if (!_assetGraph.contains(id)) return UnreadableAssetReason.notFound;
+    var node = _assetGraph.get(id);
+    if (node.isDeleted) return UnreadableAssetReason.deleted;
+    if (!node.isReadable) return UnreadableAssetReason.assetType;
+    if (node is GeneratedAssetNode) {
+      if (node.isFailure) return UnreadableAssetReason.failed;
+      if (!node.wasOutput) return UnreadableAssetReason.notOutput;
+    }
+    if (await _delegate.canRead(id)) return null;
```

Comment:
maybe throw?

Reply:
This is a strange api, `null` is actually the typical case.

Maybe I should try it more like a `Maybe` or `Optional` or something, where I add a read API that either gives you the `List<int>` or gives you an `UnreadableAssetReason`...

Reply:
Hmm, even that wouldn't accomplish what we need since we only want to get the bytes after checking the digest and deciding we have a cache miss on the browser side.

I suppose I could reintroduce a call to `canRead` and only check the reason if `canRead` returned false. I had hoped to not do all the checking twice but that might be a premature optimization.

Reply:
Ya I think just leaving in the `canRead` call and then having a separate call that expects to only be given unreadable asset ids would be a bit cleaner.